### ISSUE TITLE
Regression: False positive type error/wrong type inference for macro-defined integer methods

### ISSUE THREAD

**Issue by Ben-Kerman** [intellij-rust/intellij-rust#8969]
## Environment

* **IntelliJ Rust plugin version:** 0.4.174.4704-221-nightly (also observed on stable version 0.4.172.4687-221)
* **Rust toolchain version:** 1.61.0 (fe5b13d68 2022-05-18) x86_64-unknown-linux-gnu
* **IDE name and version:** CLion 2022.1 (CL-221.5080.224)
* **Operating system:** Linux 5.18.5-arch1-1
* **Macro expansion engine:** new
* **Name resolution engine:** new
* **Additional experimental features:** org.rust.cargo.evaluate.build.scripts, org.rust.macros.proc (but  also occurs with those features disabled)

## Problem description

Using integer methods such as saturating_add like in the example below results in a type error, even though the code compiles successfully. This issue was not yet present in release 170.

I suspect this has something to do with those methods being defined through a macro (`int_impl!` in core/num/int_macros.rs, invoked in core/num/mod.rs).

## Steps to reproduce
Create a new project and paste the following code into main.rs:
```rust
use std::collections::HashMap;

fn main() {
    let map: HashMap<u32, u16> = (0..16).fold(HashMap::new(), |mut map, c| {
        let entry = map.entry(c).or_insert(0);
        *entry = entry.saturating_add(1);
        map
    });
}
```
This results in this error message:
![image](https://user-images.githubusercontent.com/21299619/175423778-f7987d0b-d46b-4c8c-83d0-7959311cae76.png)
